protocol_decoder/src/decoding.rs

```diff
@@ -281,8 +298,74 @@ impl ProcessedBlockTrace {
                 let dummy_txn = create_dummy_gen_input(other_data, initial_trie_state, 0);
                 gen_inputs.insert(0, dummy_txn);
             }
-            _ => (),
+            _ => dummies_added = false,
         }
+
+        dummies_added
+    }
+
+    /// The withdrawals are always in the final ir payload. How they are placed
+    /// differs based on whether or not there are already dummy proofs present
+    /// in the IR. The rules for adding withdrawals to the IR list are:
+    /// - If dummy proofs are already present, then the withdrawals are added to
+    ///   the last dummy proof (always index `1`).
+    /// - If no dummy proofs are already present, then a dummy proof that just
+    ///   contains the withdrawals is appended to the end of the IR vec.
+    fn add_withdrawals_to_txns(
+        txn_ir: &mut Vec<TxnProofGenIR>,
+        other_data: &OtherBlockData,
+        final_trie_state: &mut PartialTrieState,
+        withdrawals: Vec<(Address, U256)>,
+        dummies_already_added: bool,
+    ) -> TraceParsingResult<()> {
+        // Withdrawals update balances in the account trie, so we need to do that here.
+        Self::update_trie_state_from_withdrawals(withdrawals.iter(), &mut final_trie_state.state)?;
```

Comment:
How about making `update_trie_state_from_withdrawals` take `impl IntoIterator<… …>` instead? Sp you can avoid this `.iter()`?